# Project 3

In this project, you will perform a logistic regression on the admissions data we've been working with in projects 1 and 2.

In [23]:
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.api as sm
import pylab as pl
import numpy as np
%matplotlib inline

In [4]:
df = pd.read_csv("../assets/admissions.csv")
df = df.dropna() 
print(df.head())

   admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0


## Part 1. Frequency Tables

#### 1. Let's create a frequency table of our variables

In [5]:
# frequency table for prestige and whether or not someone was admitted
df.groupby('prestige')['admit'].value_counts()

prestige  admit
1.0       1        33
          0        28
2.0       0        95
          1        53
3.0       0        93
          1        28
4.0       0        55
          1        12
Name: admit, dtype: int64

## Part 2. Return of dummy variables

#### 2.1 Create class or dummy variables for prestige 

In [6]:
df.loc[:,'prestige'] = df.loc[:,'prestige'].astype(int)
prestige = pd.get_dummies(df['prestige'],drop_first=False, prefix='prestige')
df = pd.concat([df, prestige], axis=1)
df.head()

admit    gre   gpa  prestige  prestige_1  prestige_2  prestige_3  \
0      0  380.0  3.61         3           0           0           1   
1      1  660.0  3.67         3           0           0           1   
2      1  800.0  4.00         1           1           0           0   
3      1  640.0  3.19         4           0           0           0   
4      0  520.0  2.93         4           0           0           0   

   prestige_4  
0           0  
1           0  
2           0  
3           1  
4           1

#### 2.2 When modeling our class variables, how many do we need? 



Answer: what is "modeling our class variables"? 

## Part 3. Hand calculating odds ratios

Develop your intuition about expected outcomes by hand calculating odds ratios.

In [10]:
# frequency table for prestige and whether or not someone was admitted
print(pd.crosstab(df['admit'],df['prestige']))

prestige   1   2   3   4
admit                   
0         28  95  93  55
1         33  53  28  12


#### 3.1 Use the cross tab above to calculate the odds of being admitted to grad school if you attended a #1 ranked college

In [13]:
odds1 = 33.0/28
odds1

1.1785714285714286

#### 3.2 Now calculate the odds of admission if you did not attend a #1 ranked college

In [14]:
odds2 = (53+28+12)/(95+93+55)
odds2

0.38271604938271603

#### 3.3 Calculate the odds ratio

In [15]:
odds_ratio = odds1/odds2
odds_ratio

3.079493087557604

#### 3.4 Write this finding in a sentenance: 

Answer: The odds ratio for one who attended a #1 ranked college to get admitted to one who did not is 3.08

#### 3.5 Print the cross tab for prestige_4

In [16]:
print(pd.crosstab(df['admit'],df['prestige'][df['prestige']==4]))

prestige  4.0
admit        
0          55
1          12


#### 3.6 Calculate the OR 

In [17]:
odds_ratios_4 = ((12/55)/((53+28+33)/(95+93+28)))
odds_ratios_4

0.4133971291866028

#### 3.7 Write this finding in a sentence

Answer: The odds ratio for one who attended a #4 ranked college to get admitted to one who did not is 0.41

## Part 4. Analysis

In [30]:
# create a clean data frame for the regression
cols_to_keep = ['admit', 'gre', 'gpa']
dummy_ranks = pd.get_dummies(df['prestige'], prefix='prestige')
data = df[cols_to_keep].join(dummy_ranks.loc[:, 'prestige_2':])
data.head()

admit    gre   gpa  prestige_2  prestige_3  prestige_4
0      0  380.0  3.61           0           1           0
1      1  660.0  3.67           0           1           0
2      1  800.0  4.00           0           0           0
3      1  640.0  3.19           0           0           1
4      0  520.0  2.93           0           0           1

We're going to add a constant term for our Logistic Regression. The statsmodels function we're going to be using requires that intercepts/constants are specified explicitly.

In [31]:
# manually add the intercept
data['intercept'] = 1.0

#### 4.1 Set the covariates to a variable called train_cols

In [33]:
train_cols = data.columns[1:]

#### 4.2 Fit the model

In [34]:
lm = sm.Logit(data['admit'],data[train_cols]).fit()

Optimization terminated successfully.
         Current function value: 0.573854
         Iterations 6


from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_curve, roc_auc_score, accuracy_score, auc, confusion_matrix, recall_score, precision_score
%matplotlib inline
lm = LogisticRegression().fit(data[train_cols],data['admit'])

#### 4.3 Print the summary results

In [35]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                  admit   No. Observations:                  397
Model:                          Logit   Df Residuals:                      391
Method:                           MLE   Df Model:                            5
Date:                Tue, 04 Jul 2017   Pseudo R-squ.:                 0.08166
Time:                        01:03:47   Log-Likelihood:                -227.82
converged:                       True   LL-Null:                       -248.08
                                        LLR p-value:                 1.176e-07
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
gre            0.0022      0.001      2.028      0.043    7.44e-05       0.004
gpa            0.7793      0.333      2.344      0.019       0.128       1.431
prestige_2    -0.6801      0.317     -2.146      0.032      -1.301      -0.059
prestige_3    -1.3387      0.345     -3.882      0.000      -2.015      -0.663
prestige_4    -1.5534      0.417     -3.721      0.000      -2.372      -0.735
intercept     -3.8769      1.142     -3.393      0.001      -6.116      -1.638
==============================================================================
"""

#### 4.4 Calculate the odds ratios of the coeffiencents and their 95% CI intervals

hint 1: np.exp(X)

hint 2: conf['OR'] = params
        
           conf.columns = ['2.5%', '97.5%', 'OR']

$$p = \frac{e^{\beta_0 + \beta_1x}} {1 + e^{\beta_0 + \beta_1x}}$$


In [36]:
np.exp(lm.params)

gre           1.002221
gpa           2.180027
prestige_2    0.506548
prestige_3    0.262192
prestige_4    0.211525
intercept     0.020716
dtype: float64

In [38]:
conf = lm.conf_int()
conf['OR']=lm.params
conf.columns = ['2.5%', '97.5%', 'OR']
np.exp(conf)

2.5%     97.5%        OR
gre         1.000074  1.004372  1.002221
gpa         1.136120  4.183113  2.180027
prestige_2  0.272168  0.942767  0.506548
prestige_3  0.133377  0.515419  0.262192
prestige_4  0.093329  0.479411  0.211525
intercept   0.002207  0.194440  0.020716

In [ ]:
np.exp(lm.params[0]+lm.params[1])

#### 4.5 Interpret the OR of Prestige_2

Answer: 0.51

#### 4.6 Interpret the OR of GPA

Answer: 2.18

## Part 5: Predicted probablities


As a way of evaluating our classifier, we're going to recreate the dataset with every logical combination of input values. This will allow us to see how the predicted probability of admission increases/decreases across different variables. First we're going to generate the combinations using a helper function called cartesian (above).

We're going to use np.linspace to create a range of values for "gre" and "gpa". This creates a range of linearly spaced values from a specified min and maximum value--in our case just the min/max observed values.

In [39]:
def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.
    Parameters
    ----------
    arrays : list of array-like
        1-D arrays to form the cartesian product of.
    out : ndarray
        Array to place the cartesian product in.
    Returns
    -------
    out : ndarray
        2-D array of shape (M, len(arrays)) containing cartesian products
        formed of input arrays.
    Examples
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = n / arrays[0].size
    out[:,0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m,1:])
        for j in xrange(1, arrays[0].size):
            out[j*m:(j+1)*m,1:] = out[0:m,1:]
    return out

In [48]:
# instead of generating all possible values of GRE and GPA, we're going
# to use an evenly spaced range of 10 values from the min to the max 

gres = np.linspace(data['gre'].min(), data['gre'].max(), 10)
print(gres)
# array([ 220.        ,  284.44444444,  348.88888889,  413.33333333,
#         477.77777778,  542.22222222,  606.66666667,  671.11111111,
#         735.55555556,  800.        ])
gpas = np.linspace(data['gpa'].min(), data['gpa'].max(), 10)
print(gpas)
# array([ 2.26      ,  2.45333333,  2.64666667,  2.84      ,  3.03333333,
#         3.22666667,  3.42      ,  3.61333333,  3.80666667,  4.        ])


# enumerate all possibilities
combos = pd.DataFrame(cartesian([gres, gpas, [1, 2, 3, 4], [1.]]))
# recreate the dummy variables


[ 220.          284.44444444  348.88888889  413.33333333  477.77777778
  542.22222222  606.66666667  671.11111111  735.55555556  800.        ]
[ 2.26        2.45333333  2.64666667  2.84        3.03333333  3.22666667
  3.42        3.61333333  3.80666667  4.        ]


TypeError: slice indices must be integers or None or have an __index__ method

#### 5.1 Recreate the dummy variables

In [ ]:
# recreate the dummy variables
combos.columns = ['gre', 'gpa', 'prestige', 'intercept']
dummy_ranks = pd.get_dummies(combos['prestige'], prefix='prestige')
dummy_ranks.columns = ['prestige_1', 'prestige_2', 'prestige_3', 'prestige_4']

# keep only what we need for making predictions
cols_to_keep = ['gre', 'gpa', 'prestige', 'intercept']
combos = combos[cols_to_keep].join(dummy_ranks.ix[:, 'prestige_2':])


#### 5.2 Make predictions on the enumerated dataset

#### 5.3 Interpret findings for the last 4 observations

Answer: 

## Bonus

Plot the probability of being admitted into graduate school, stratified by GPA and GRE score.